### DISCUSSION TITLE
zot4rst + zotxt equivalent for papis?

### DISCUSSION THREAD

**Discussion by mcepl** [papis/papis#492] [Ideas]
Currently I use [reStructuredText](https://docutils.sourceforge.io/rst.html) together with [zot4rst](https://github.com/egh/zot4rst) and [zotxt](https://gitlab.com/egh/zotxt) to write my blog posts. For an example see [this rST source](https://git.cepl.eu/cgit/websites/blog-source/tree/computer/vim-acme-design.rst) and [its presentation on the blog](https://matej.ceplovi.cz/blog/design-notes-on-plan9-for-vimspace.html) or [another example](https://git.cepl.eu/cgit/websites/blog-source/tree/faith/greek_church_orthodoxy.rst) (and [its HTML rendering](https://matej.ceplovi.cz/blog/greek-church-orthodoxy.html)).

It works, but Zotero is awfully giant monster which must run in the background whenever I [update my blogs](https://getpelican.com/) without regards whether in the moment I use it or not. Therefore, I would like to switch to something more lightweight and papis seems to fit the bill. Any ideas for some support of using papis with reStructuredText?

**Comment by alexfikl** [comment]
Hey, thanks for the interest in papis!

At least in core `papis`, there's no support at the moment for some of the things that `zot4rst` and `zotxt` seem to do. From a quick skim, we'd need some plugins for `sphinx`, `pelican` and `pandoc` to make it all work. Hopefully those are just some simple wrappers around core `papis` functionality.

The big chunk that seems to be missing is support for citation styles, but maybe [citeproc-py](https://github.com/brechtm/citeproc-py) can be used here without much hassle.

I don't know if anyone has plans to implement some of that, but we're more than happy to help if you get to it! @alejandrogallo probably has a better view of the ecosystem, since I mostly keep to straight up LaTeX + BibTeX.

**Comment by alejandrogallo** [comment]
I think in principle it should be quite straight forward-ish to adapt
the codebase of zot4rst to support papis, using citeproc-py as Alex
discussed. However I am quite surprised that there is no
easy way to use bibtex with rst, this would solve much of the problems
that @mcepl is having, that way you would basically be independent.

Personally what I would do is do this in a script,
maybe something like here

https://github.com/cykustcc/bib2reSTcitation/blob/master/references.txt

But it does not have to be so difficult. For instance in my website I have this
template

https://github.com/alejandrogallo/alejandrogallo.github.io/blob/master/publication.template

```
1. *{doc[title]}* \\
  /{doc[author]}/\\
  *{doc[year]}*, {doc[publisher]}
  [[https://doi.org/{doc[doi]}][doi]]
```

that I creates org mode items,

and I use this targets in the Makefile

```make

publications.org: publication.template
	papis list \
		--sort year \
		--reverse \
		--template publication.template \
		--all gallo:true > $@
	sed -i "s/Gallo, Alejandro/*Gallo, Alejandro*/" publications.org
	sed -i "s/Alejandro Gallo/*Alejandro Gallo*/" publications.org
```

where I have in my publications in papis the field

```yaml
gallo: true
```

You could do something like this, where you create the
reference rst file bit by bit using a template, and then
just referencing the entries in the bib rst file.

This would be a workaround for simple stuff, which would
work for both latex and html exporting.

**Reply by mcepl** [reply]
Well, I didn’t want to use BibTeX as in-between format, every conversion looses information and control.

Hmm, is there some support for [CSL](https://citationstyles.org/) in papis? That could be useful.

**Reply by alejandrogallo** [reply]
there is no CSL support at the moment in papis, but I can see it might be useful for people like yourself and maybe also in the future for possible other extensions. I could implement it using citeproc-py, it seems a quite reasonable library.

What do you think @alexfikl? Should we put it in core or as a plugin? What about the cli ? 

Should it be something like

```
papis export --csl harvard1 author:einstein
```

or rather 
```
papis export --format csl:harvard
```
?

The latter would be a little more difficult to implement, the former quite starightforward

**Reply by alejandrogallo** [reply]
I am having quite a lot of trouble with https://github.com/brechtm/citeproc-py, I also don't really like that they wrote their own parser for bibtex which complains for me about keys like `issue`, `eprint`, `date`, `url`, `school`, `language` and some others. I'll try their JSON importer maybe, does someone incidentally know of any other library implementing CSL ?

**Reply by alejandrogallo** [reply]
On a side note, they're looking for maintainers, so maybe the papis community could take care of this package like we did with python-dmenu stuff or python-doi

**Reply by alexfikl** [reply]
At least for the command-line interface, I think the simplest option would probably be just the usual
```
papis export --format csl
```
and have the style as a config option, e.g. `csl-style-name`? That wouldn't require adding any new options to the `export` cli or changing the API for the exporters `export(docs: List[Document])`. What do you think?

As for `citeproc-py`, I just found the library from some googling and don't have any experience with it. I can't seem to find any other libraries for CSL in python, so we're probably stuck with it :\

**Reply by mcepl** [reply]
There is https://pypi.org/search/?q=CSL , but I am afraid `citeproc-py` is truly the official module for CSL.

**Comment by arinbasu** [comment]
Hi Maintainers,
This is a fantastic package, I accidentally stumbled upon it to replace my JabRef (this was not working). On an Archlinux installation, `papis open` does not seem to work as expected as it opens the PDF in the terminal window itself. I have indicated on the config file that it should open either in evince or xdg-open but neither seem to work. `papis serve` works but only if we type 127.0.0.1:8888 (this could be a computer issue) but it seems to work.
Excellent work! I think there are many use cases now, if we can integrate this with other tools such as JupyterLab etc. is it available as a python package, so can we do something like `import papis` to manage bibliography in Python?
Best,
Arin

**Reply by alexfikl** [reply]
Thank you for the kind words!

> On an Archlinux installation, `papis open` does not seem to work as expected as it opens the PDF in the terminal window itself. I have indicated on the config file that it should open either in evince or xdg-open but neither seem to work.

I'm also on Arch and this seems to work correctly for me (`xdg-open` is the default value for `opentool`). Can you share what commands you've tried running and what output they gave in more detail? Maybe the output of something like
```
papis --log DEBUG open <query>
```

> `papis serve` works but only if we type 127.0.0.1:8888 (this could be a computer issue) but it seems to work.

That's the correct behaviour. You can pick the address and port to bind to when calling `papis server` using
```
papis serve --address localhost --port 8888
```
if those don't work for you. What behaviour did you expect here?

> is it available as a python package, so can we do something like import papis to manage bibliography in Python?

Yes, definitely, you can import `papis` and use its functionality to make your own plugins and scripts as you like -- it's generally very flexible in that regard. Some places you can look at are
* The official API (that we try to keep stable) is documented at https://papis.readthedocs.io/en/latest/api.html. We're trying to add documentation for other parts of `papis` at https://papis.readthedocs.io/en/latest/developer_reference.html, but these are generally internal APIs so use at your own risk.
* There's some docs on how to add plugins at https://papis.readthedocs.io/en/latest/plugins.html. Plugins for commands, importers, downloaders, pickers, etc are all possible, so you can change most parts of `papis` that you don't like.
* There's some example command plugins at https://github.com/papis/papis/tree/main/examples/scripts

Hope that helps!

**Reply by alexfikl** [reply]
By the way, feel free to open some separate issues for the problems you're having! There may be some actual bugs in there!
